### PR TITLE
iteration documentation and limitations based on unit tests first draft

### PR DESCRIPTION
**Pull request by moAlobaidi**
Hey @collinc97 , here's a first draft of the documentation of loops and iteration in Leo based on the unit tests you gave me for reference. Lmk if any of this information is not accurate or if there is anything else I should add! Also feel free to give any feedback necessary on the style / wording. 

This PR resolves issue #56

### CODE REVIEW

File: documentation/leo/03_language.md [hunk 1]
```diff
@@ -412,23 +412,127 @@ program test.aleo {
...
+```
+
+The end bound is too large to be represented as a `u64`.
+
+## Notes
+
+1. Leo doesn't currently support the `break` and `continue` statements found in other languages such as
+   Rust. These statements allow for more complex control flow within loops, but are not currently
+   available in Leo.
+
```
Comment (collinc97) [inline documentation/leo/03_language.md:535]: make these more specific: say exactly what the minimum and maximum loop requirements are or remove these docs

**Review by collinc97 — CHANGES_REQUESTED** [review]
Thanks for the PR @moAlobaidi!
Definitely looks like you spent some time playing around with loops. An overall comment from me would be to follow the length and diction of the other Leo docs.

File: documentation/leo/03_language.md [hunk 2]
```diff
@@ -412,23 +412,127 @@ program test.aleo {
...
+```leo
+for i: u32 in 0u32..9u32 {
+    return false;
+}
+```
+
+The `return` statement is inside the loop, which is not allowed in Leo.
+
+## Loop Errors
```
[the remark this answers is not in the document]
Comment (moAlobaidi) [reply documentation/leo/03_language.md:496]: Good point! Would an alternate implementation for non-literal loop bounds be asserting that the bound isn't too large?

File: documentation/leo/03_language.md [hunk 3]
```diff
@@ -412,106 +412,48 @@ program test.aleo {
...
-
-In the Leo programming language, iteration is supported through the `for` loop construct. This section provides an
-overview of how
-iteration works in Leo and highlights its limitations.
-
-### `for` Loop Syntax
+## For Loops
 
-The syntax for a `for` loop in Leo is as follows:
+For Loops are declared as `for <loop variable> in <lower bound>..<upper bound>`.  Currently, the
```
Comment (aharshbe) [inline documentation/leo/03_language.md:435]: For consistency with the existing docs, let's change <> to {}, so `for {variable} in {lower-bound}..{upper-bound}`. A good example of this to reference is the [mapping type](https://developer.aleo.org/leo/language#mapping).

File: documentation/leo/03_language.md [hunk 4]
```diff
@@ -412,106 +412,48 @@ program test.aleo {
...
+loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be
+less than the upper bound.
 
-```
+```leo
 for <loop variable> in <lower bound>..<upper bound> {
     // Loop body
```
Comment (aharshbe) [inline documentation/leo/03_language.md:441]: Same thing here change <> to {}

File: documentation/leo/03_language.md [hunk 5]
```diff
@@ -412,106 +412,48 @@ program test.aleo {
...
-overview of how
-iteration works in Leo and highlights its limitations.
-
-### `for` Loop Syntax
+## For Loops
 
-The syntax for a `for` loop in Leo is as follows:
+For Loops are declared as `for <loop variable> in <lower bound>..<upper bound>`.  Currently, the
+loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be
+less than the upper bound.
```
Comment (aharshbe) [inline documentation/leo/03_language.md:437]: After this sentence, add "Nested loops are supported." and remove the example of a nested loop below.

File: documentation/leo/03_language.md [hunk 6]
```diff
@@ -412,106 +412,48 @@ program test.aleo {
...
 }
 ```
 
-The `for` loop starts by declaring a loop variable that will take on different values in each iteration. The loop
-variable currently supports the `u8`, `u16`, or `u32` data types. It is then
-followed by the `in` keyword and a range specified by the lower and upper bounds.
+### Nested Loops
 
-### Example:
+You can nest loops within loops. For example:
```
Comment (aharshbe) [inline documentation/leo/03_language.md:447]: I don't think you need to provide an example here. Just say nested loops are supported.

**Review by aharshbe — CHANGES_REQUESTED** [review]
I left a few simple changes here. Happy to make them myself but want to make sure you have some eyes on it first @moAlobaidi 🙂 Nice work!

**Comment by aharshbe** [comment]
Once these changes are complete, this is good to merge.

**Comment by aharshbe** [comment]
Let's hold on merging this until we confirm u64 vs u32 with Pranav tomorrow.

File: documentation/leo/03_language.md [hunk 7]
```diff
@@ -423,12 +424,24 @@ If the value at `addr` does not exist, then the program will fail to execute.
...
+If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.
 
 #### set
 A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
 Sets the `addr` entry as `current_value + 1u64` in `counter`.
 
+## For Loops
+
+For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Currently, the
+loop variable supports unsigned integer types u8, u16, u32, or u64. The lower bound must be
```
Comment (aharshbe) [inline documentation/leo/03_language.md:436]: @moAlobaidi, this should change to support up to u32. Just got confirmation from the Leo team. For context, internally we support up to u128 but externally we shouldn't mention this.

**Review by aharshbe — APPROVED** [review]
@moAlobaidi, we just need to update to be up to u32 so remove u64. Going to approve this so once you made those changes feel free to merge!

cc @collinc97

**Review by aharshbe — APPROVED** [review]

File: documentation/leo/03_language.md (hunk 7 above)
Comment (collinc97) [inline documentation/leo/03_language.md:436]:
```suggestion
For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Typical Leo loop variable types include `u8`, `u16`, and `u32`. The lower bound must be
```

Comment (collinc97) [reply documentation/leo/03_language.md:436]: What is a `{loop variable}` it is not defined in the documentation.

File: documentation/leo/03_language.md [hunk 8]
```diff
@@ -423,12 +424,24 @@ If the value at `addr` does not exist, then the program will fail to execute.
...
+less than the upper bound. Nested loops are supported.
+
+```leo
+for {loop variable} in {lower bound}..{upper bound} {
+    // Loop body
+}
+```
+
```
Comment (collinc97) [inline documentation/leo/03_language.md:444]: You should add a code snippet of a loop that actually compiles.
People will want to copy and paste code from the docs into their program to play around with the feature.

**Review by collinc97 — CHANGES_REQUESTED** [review]

File: documentation/leo/03_language.md (hunk 7 above)
Comment (collinc97) [reply documentation/leo/03_language.md]: That description still does not help the programmer write a loop @moAlobaidi.
Loops in Leo require a variable name and a variable type.
`{loop variable}` should be defined as `{name: type}` with an example.

Comment (moAlobaidi) [reply documentation/leo/03_language.md]: Okay updated.

File: documentation/leo/03_language.md [hunk 9]
```diff
@@ -412,23 +432,47 @@ program test.aleo {
...
+If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.
 
 #### set
+
 A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
 Sets the `addr` entry as `current_value + 1u64` in `counter`.
 
+## For Loops
+
+For Loops are declared as `for {variable_name: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer
```
Comment (collinc97) [inline documentation/leo/03_language.md:459]:
```suggestion
For Loops are declared as `for {variable: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 
```
Remove mixed usage of snake_case and text in description

File: documentation/leo/03_language.md [hunk 10]
```diff
@@ -412,23 +432,47 @@ program test.aleo {
...
 #### set
+
 A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
 Sets the `addr` entry as `current_value + 1u64` in `counter`.
 
+## For Loops
+
+For Loops are declared as `for {variable_name: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 
+types of up to
+u32 are recommended for iterator variables. The lower bound must be
```
Comment (collinc97) [inline documentation/leo/03_language.md:461]:
```suggestion
types `u8`, `u16`, and `u32` are recommended for loop variable types. The lower bound must be
```
* State explicit types.
* Add backticks to types so they show up as `code` - this is consistent with the rest of `u8` mentions throughout the documentation.
* Remove usage of "iterator" name which has not been mentioned before in documentation
* Specify that we are describing `type` of. the iterator variable.

File: documentation/leo/03_language.md [hunk 11]
```diff
@@ -412,23 +432,47 @@ program test.aleo {
...
+less than the upper bound. Nested loops are supported.
+
+### Example
+
+```leo
+  let count: u32 = 0u32;
+
+  for i:u32 in 0u32..5u32 {
```
Comment (collinc97) [inline documentation/leo/03_language.md:469]:
```suggestion
  for i: u32 in 0u32..5u32 {
```

**Review by collinc97 — CHANGES_REQUESTED** [review]
final changes

File: documentation/leo/03_language.md [hunk 12]
```diff
@@ -412,23 +432,46 @@ program test.aleo {
...
+If the key is not present, `0u64` is stored in `counter` and stored in `current_value`.
 
 #### set
+
 A set command, e.g. `Mapping::set(counter, addr, current_value + 1u64);`
 Sets the `addr` entry as `current_value + 1u64` in `counter`.
 
+## For Loops
+
+For Loops are declared as `for {variable: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer
```
Comment (collinc97) [inline documentation/leo/03_language.md:459]:
```suggestion
For Loops are declared as `for {variable: type} in {lower bound}..{upper bound}`. Unsigned integer 
```
The lower and upper bounds are literals - their type is specified with the value `1u8`, `1337u16`
There is no colon required @moAlobaidi

**Review by collinc97 — CHANGES_REQUESTED** [review]

File: documentation/leo/03_language.md [hunk 13]
```diff
@@ -412,23 +432,46 @@ program test.aleo {
...
+### Example
+
+```leo
+  let count: u32 = 0u32;
+
+  for i: u32 in 0u32..5u32 {
+    count += 1u32;
+  }
+
+  return count;
```
Comment (collinc97) [inline documentation/leo/03_language.md:472]:
```suggestion
  return count; // returns 5u32
```

**Pull request merged**